File: piro-api/backend/db/models/CytologyEvaluation.py
from db.base_class import Base
from sqlalchemy import Boolean, Column, DateTime, ForeignKey, Integer, String
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func


class CytologyEvaluation(Base):
    __tablename__ = "CytologyEvaluation"
    CytologyEvaluationId = Column(Integer, primary_key=True, index=True)
    PatientIdentifiers = Column(String, nullable=True)
    ProcedureType = Column(String, nullable=True)
    ProcedurePerformedBy = Column(String, nullable=True)
    EvaluationPerformedBy = Column(String, nullable=True)
    ViaTelecytology = Column(Boolean, nullable=True)
    ReadLocation = Column(String, nullable=True)
    ProcedureLocation = Column(String, nullable=True)
    AssignedToUserId = Column(Integer, ForeignKey("User.UserId"), nullable=True)
    ClinicalHistory = Column(String, nullable=True)
    Notes = Column(String, nullable=True)
    PatientHistory = Column(String, nullable=True)
    CytologyPersonnelUserId = Column(
        Integer, ForeignKey("User.UserId"), nullable=True
    )
    PathologistUserId = Column(Integer, ForeignKey("User.UserId"), nullable=True)
    FellowUserId = Column(Integer, ForeignKey("User.UserId"), nullable=True)
    ResidentUserId = Column(Integer, ForeignKey("User.UserId"), nullable=True)
    TotalTimeSpentMinutes = Column(Integer, nullable=True)
    Status = Column(String, nullable=False, default="Draft")
    PrelimVerifierId = Column(Integer, ForeignKey("User.UserId"), nullable=True)
    PrelimVerifiedDate = Column(DateTime(timezone=False), nullable=True)
    FinalVerifierId = Column(Integer, ForeignKey("User.UserId"), nullable=True)
    FinalVerifiedDate = Column(DateTime(timezone=False), nullable=True)
    # Use timezone-naive datetimes to match the SQL schema (datetime, not datetimeoffset)
    CreateDate = Column(DateTime(timezone=False), default=func.now())
    CreateBy = Column(String, nullable=False)
    UpdateDate = Column(DateTime(timezone=False), onupdate=func.now())
    UpdateBy = Column(String, nullable=True)

    AssignedTo = relationship("User", foreign_keys=[AssignedToUserId])
    CytologyPersonnel = relationship(
        "User", foreign_keys=[CytologyPersonnelUserId]
    )
    Pathologist = relationship("User", foreign_keys=[PathologistUserId])
    Fellow = relationship("User", foreign_keys=[FellowUserId])
    Resident = relationship("User", foreign_keys=[ResidentUserId])
    PrelimVerifier = relationship("User", foreign_keys=[PrelimVerifierId])
    FinalVerifier = relationship("User", foreign_keys=[FinalVerifierId])
    Sites = relationship(
        "CytologyEvaluationSite",
        back_populates="CytologyEvaluation",
        order_by="CytologyEvaluationSite.SortOrder",
        cascade="all, delete-orphan",
    )

File: piro-ui/src/app/models/cytology-evaluation.ts
export type CytologyEvaluationStatus = 'Draft' | 'Prelim Verified' | 'Final Verified';

export interface CytologyEvaluationSite {
  id: number;
  site?: string | null;
  evalEpisodeNumber?: number | null;
  adequacy?: string | null;
  dqCount: number;
  papCount: number;
  thinPrepCount: number;
  cellBlockCount: number;
  unstainedSlidesCount: number;
  sortOrder: number;
}

export interface CytologyEvaluationSiteInput {
  id?: number | null;
  site?: string | null;
  evalEpisodeNumber?: number | null;
  adequacy?: string | null;
  dqCount: number;
  papCount: number;
  thinPrepCount: number;
  cellBlockCount: number;
  unstainedSlidesCount: number;
}

export interface CytologyEvaluationTotals {
  totalDQ: number;
  totalPap: number;
  totalThinPrep: number;
  totalCellBlock: number;
  totalUnstainedSlides: number;
}

export interface CytologyEvaluationSavePayload {
  patientIdentifiers?: string | null;
  procedureType?: string | null;
  procedurePerformedBy?: string | null;
  evaluationPerformedBy?: string | null;
  viaTelecytology?: boolean | null;
  readLocation?: string | null;
  procedureLocation?: string | null;
  assignedToUserId?: number | null;
  clinicalHistory?: string | null;
  notes?: string | null;
  patientHistory?: string | null;
  cytologyPersonnelUserId?: number | null;
  pathologistUserId?: number | null;
  fellowUserId?: number | null;
  residentUserId?: number | null;
  totalTimeSpentMinutes?: number | null;
  sites: CytologyEvaluationSiteInput[];
}

export interface CytologyEvaluation extends CytologyEvaluationSavePayload {
  id: number;
  status: CytologyEvaluationStatus;
  assignedToName?: string | null;
  cytologyPersonnelName?: string | null;
  pathologistName?: string | null;
  fellowName?: string | null;
  residentName?: string | null;
  prelimVerifierNuid?: string | null;
  prelimVerifierName?: string | null;
  prelimVerifiedDate?: string | null;
  finalVerifierNuid?: string | null;
  finalVerifierName?: string | null;
  finalVerifiedDate?: string | null;
  createDate: string;
  updateDate?: string | null;
  sites: CytologyEvaluationSite[];
  totals: CytologyEvaluationTotals;
}

export interface CytologyTerminology {
  procedureType: string[];
  readLocation: string[];
  procedureLocation: string[];
  site: string[];
  adequacy: string[];
}

export interface UserSearchResult {
  userId: number;
  nuid: string;
  firstName: string;
  lastName: string;
}
